### DISCUSSION TITLE
Can't run old Algorand devnet anymore

### DISCUSSION THREAD

**Discussion by DanBurton** [reach-sh/reach-lang#1290] [Q&A] [answered]
I used to be able to run this:

```
$ REACH_VERSION=2604daf0 REACH_CONNECTOR_MODE=ALGO reach devnet
```

And it would work fine. But at some point I did a `reach update` and now it doesn't work, even though I'm specifying the same (old) version as before. It just says this over and over again:

```
/start.sh: line 11: : Name does not resolve
/start.sh: line 11: /dev/tcp//5432: Invalid argument
not ready yet, trying again in 1s...
```

**Comment by DanBurton** [comment, marked as answer]
# Context

There was a change in Reach's algorand devnet where we moved the postgres db inside of the devnet Docker image, instead of relying on it as a separate Docker image. Unfortunately, this has hampered one's ability to run old devnets.

There are three versioned things to understand about what's going on here:

1. the version of one's `reach` script
2. the version of one's `reachsh/reach-cli` Docker image
3. the version of one's `reachsh/devnet-algo` Docker image

For #1, you can pull an old version of the `reach` script directly off of github, though this is likely unnecessary. See, for example: https://github.com/reach-sh/reach-lang/blob/2604daf0/reach

As for #s 2 and 3, you can see their version by calling `reach hashes`.

```
$ REACH_VERSION=2604daf0 reach hashes
reach-cli: a9f7613d
...
devnet-algo: 2604daf0
```

Notice how, for me, `reach-cli` did not indicate the same hash. That is because `reach-cli` always uses its `:latest` version.

# Solutions

There are two ways to work around this issue:

## just run a newer version of the devnet

Newer versions of the Algorand devnet may still be compatible with your smart contracts, even if they were compiled with an older `reach` version.

## retag `reachsh/reach-cli:latest` to use an older version

You can manually retag `reach-cli:latest` to force it to use its older version like so:

```
REACH_VERSION=2604daf0 # or whatever your desired version is
docker pull "reachsh/reach-cli:$REACH_VERSION"
docker tag "reachsh/reach-cli:$REACH_VERSION" reachsh/reach-cli:latest
```

(This will be undone if you `reach update`)

**Reply by jeapostrophe** [reply]
"may still be compatible" --- Algorand has never implemented a backwards incompatible change and doesn't intend to. New devotes should _always_ be compatible or it will be very very bad for Algorand.

**Reply by DanBurton** [reply]
Tangential but related: we at Reach are reworking the Reach command-line application in a way that will avoid this kind of problem in future versions.

**Reply by DanBurton** [reply]
@jeapostrophe I agree.

However, the reach-provided devnets of a given version also have parity with the Reach stdlib's `newTestAccount` and `fundFromFaucet` functionality (when the stdlib version matches). Sometimes the faucet account gets changed across versions, which causes backwards-incompatibilities only in regards to this particular faucet-related functionality.

But even in this case, one can use the stdlib's `setFaucet` functionality in order to override the old stdlib's expected faucet with the newer devnet's actual faucet.
